### ISSUE TITLE
X-Forwarded-Proto can be comma-delimited, but router doesn't account for that

### ISSUE THREAD

**Issue by krancour** [deis/deis#4997]
This issue supersedes #4504

**TL;DR:** I'm pretty sure we should account for the possibility of multi-valued `X-Forwarded-Proto` headers.

#4504 could not be reproduced for the longest time.  @gerred and I have finally managed to do so, and in the course of doing so, uncovered the root cause.

First a sordid little tale of service discovery:  In Deis v1.x, there is no mechanism for applications (in a micro-service architecture, perhaps) deployed to Deis to discover one another and interact directly.  So in rare cases, some users have deployed apps that talk to one another by routing traffic from one, out to the internet and back into the cluster through the load balancer and the router.  It's not ideal, but it's a workaround real users have used.

Back to our regularly scheduled bug report... Consider the following (actual) scenario: a service proxy deployed to Deis is a facade over numerous other back-end services.  It's implemented using [hapi](http://hapijs.com/).  In this example, hapi is configured to set `X-Forwarded-For` and other, related headers on the outbound HTTPS request.  These headers were already set by the Deis router on the request that was inbound to the service proxy.  hapi copies these headers over from the inbound request to the outbound request, _but also amends them_.  Where the incoming `X-Forwarded-Proto` header may have read `https`, the outbound `X-Forwarded-Proto` header will read `https,https`.

The `X-Forwarded-Proto` header, like it's better known brother `X-Forwarded-For`, is a de-facto header and is not formally described anywhere.  A brief google search has revealed however, that apart from hapi, numerous other frameworks accept or rely on the possibility that the `X-Forwarded-Proto` header might contain a comma-delimited list.  The Deis router is not one of these...

The Deis router, internally, derives a variable named `$access_scheme` from the `X-Forwarded-Proto` header, as depicted here:

```
# trust http_x_forwarded_proto headers correctly indicate ssl offloading
map $http_x_forwarded_proto $access_scheme {
    default $http_x_forwarded_proto;
    ''      $scheme;
}
```

The ultimate intention of this code is to allow the `$access_scheme` variable to assume the value `https`, even if the protocol used for the inbound request was HTTP, _as long as it appears that SSL offloading has already been performed_ (e.g. by the load balancer).

If it is possible for `X-Forwarded-Proto` to bear comma-delimited values, it is also possible for the `$access_scheme` variable to obtain a value such as `https,https`.

When Deis router's `enforceHTTPS` feature is enabled, the `$access_scheme` variable is evaluated in determining if/when a redirect to an insecure URL's secure equivalent should be performed:

```
if ($access_scheme != "https") {
    return 301 https://$host$request_uri;
}
```

In a case such as this, the value `https,https` != `https`, and as a result a redirect is issued... but this will _potentially_ initiate a series of infinite redirects.  In the specific example given, hapi was configured to follow no more than five redirects, so after the sixth, hapi quit and returned a `502` response to the router, which propagated that response back to the client.

Correcting this bug may call for some careful thought.  Some questions:

1. Should SSL offloading be considered to have occurred (the `$access_scheme` variable thereby receiving a value of `https`) if `https` occurs _anywhere_ in the comma-delimited value of the `X-Forwarded-Proto` header?  Or only if the proximate forwarded protocol was `https`?

2. Is the Deis router also guilty of _setting_ this header incorrectly on outbound requests to back-end services?  For instance, if SSL were offloaded by the load balancer, and the request inbound to the router therefore already had the `X-Forwarded-Proto` header with a value of `https`, should the router merely copy that header to the outbound request as it does today?  Or should it amend it as hapi does?  i.e. Should the outbound request have the `X-Forwarded-Proto` header set to `https,https`.  (fwiw, I believe `https,http` would be _incorrect_ since the request was treated by the router _as if_ it were secure.  This seems to be how hapi deals with this as well.)  If we fix this, do we need to / should we _also_ change how we write other `X-Forwarded-*` headers?  `X-Forwarded-For`, for instance, can be comma-delimited as well.

3. Is there an edge case where a request arrives via HTTPS, with an `X-Forwarded-Proto` header of `http`?  Such a case is possible if a client (proxy of some sort) has different opinions than us on how `X-Forwarded-Proto` should be used.  This will trip up our _existing_ logic, because such a request will be treated as insecure, despite having arrived over a secure channel... and that will _also_ trigger the infinite loop described by this issue.

4. Are there technical constraints to correcting this?  Nginx configuration isn't an imperative language, so short of resorting to the Lua module, there are limited options for tackling this.

Whatever is done, the change should also be ported forward to the Deis v2 Router.

**Comment by gerred** [comment]
Some responses to your questions:

1. I don't have an answer here. This is tough. I lean toward the latter, because it's hard to know what the trust levels of the whole chain looks like from that picture alone.

2. I'd continue setting the header as is. As of now, we've only found two frameworks that do this, and I wouldn't want to break the functionality others might be expecting / assuming, even if it's the wrong assumption. Or it's configurable. We're in Lua plugin territory then, however.

3. No idea. Can't think of a case offhand, so I'll refrain from answering here.

4. Lua is built into nginx, so I think it's fairly reasonable to add plugins if we need them. [Kong](getkong.org) is a great example of an entirely self-contained API Gateway with nginx + lua.

**Comment by krancour** [comment]
@gerred a few responses:

1. In offline discussion with you and @helgi, it seems the favored approach is to only consider the proximate protocol.  So we can try to match a regex that looks something like this: `^(?:.*,)?https$`

2. > As of now, we've only found two frameworks that do this

  I've found more mentions of comma delimited `X-Forwarded-Proto` since our initial discussion.  Having said that, I think amending these headers is one of the harder aspects of this problem.  I think I should attempt this, but not burn too much time on it.

3. If the actual protocol on an inbound request _is_ HTTPS, I think that should trump whatever the `X-Forwarded-Proto` header says when it comes to determining a value of the `$access_scheme` variable-- which ultimately dictates whether the current request is treated as secure or not.  This sounds reasonable, but it's not what happens today.  Technically, this may be a _separate_ bug, but if the logic that determines the value of `$access_scheme` is about to be re-written, it just makes sense to squash these two bugs together.

4. I want to try and avoid Lua.  If it's unavoidable, so be it, but it's possible we can achieve "good enough" without it.  No. 2 above is the bit that was most likely to have needed it, and that's the bit I'm most likely to not mess with if it gives me too much grief.

**Comment by krancour** [comment]
Having now put a little more thought into this, I think no. 2 (appending to the end of the header on the outbound request) is more important than I realized initially-- not as much on account of the `X-Forwarded-Proto` header as on the account of the complementary `X-Forwarded-For` header.  The reason for this is that in the use case I described above, the router overwriting that header instead of appending to it results in the back-end service having no notion whatsoever of the end-client IP.  Usually that information would be captured in the `X-Forwarded-For` header, but if it's overwritten outright on the second trip through the router, the IP found in that header will be the IP of the service proxy-- which is less useful information than the end-client IP.

Long story short... we _do_ / _should_ figure out the proper way (even if it involves Lua) of amending `X-Forwarded-*` headers instead of overwriting them.

**Comment by krancour** [comment]
As it turns out, the proxy_pass module has a `$proxy_add_x_forwarded_for` variable built in that is the contents of the incoming `X-Forwarded-For` header + the IP of the current remote client.  Unfortunately, no such convenience exists for the `X-Forwarded-Proto` header.

**Comment by krancour** [comment]
Ok... so we were actually already using `$proxy_add_x_forwarded_for`...

So I think that addressing nos. 1 and 3 in the list above will be sufficient.
